Modules/_uuidmodule.c

```diff
@@ -91,45 +111,1649 @@ py_windows_has_stable_node(void)
+                "bytes is not a 16-char string"
+            );
+            return -1;
+        }
+        memcpy(self->bytes, bytes->buf, 16);
+    }
+    else if (bytes_le->obj != NULL) {
+        if (from_bytes_le(self, bytes_le) < 0) {
+            return -1;
+        }
+    }
+    else if (fields != NULL) {
+        if (from_fields(self, fields) < 0) {
+            return -1;
+        }
+    }
+    else if (int_value != NULL) {
+        if (from_int(self, int_value, 1) < 0) {
+            return -1;
+        }
+    }
+    else {
+        Py_UNREACHABLE();
+    }
+
+    if (version != NULL && version != Py_None) {
+        long version_num = PyLong_AsLong(version);
+        if (version_num == -1 && PyErr_Occurred()) {
+            return -1;
+        }
+        if (version_num < 1 || version_num > 8) {
+            PyErr_SetString(PyExc_ValueError, "illegal version number");
+            return -1;
+        }
+
+        // Clear variant bits (keep only lower 6 bits of byte 8)
+        self->bytes[8] &= 0x3f;  // 0011 1111
+        // Clear version bits (keep only lower 4 bits of byte 6)
+        self->bytes[6] &= 0x0f;  // 0000 1111
+        // Set the variant to RFC 4122/9562 (binary 10xx xxxx)
+        self->bytes[8] |= 0x80;  // 1000 0000
+        // Set the version number (upper 4 bits of byte 6)
+        self->bytes[6] |= (version_num << 4);
+    }
+
+    if (is_safe != NULL) {
+        // Validate by calling SafeUUID(is_safe) to ensure it's a valid enum member
+        PyObject *validated = PyObject_CallOneArg(state->safe_uuid, is_safe);
+        if (validated == NULL) {
+            return -1;
+        }
+        Py_CLEAR(self->is_safe);
+        self->is_safe = validated;  // reuse reference
+    }
+
+    return 0;
+}
+
+
+static const uint8_t INT_TO_HEX[] = "0123456789abcdef";
+
+// Lookup table for hex character to value conversion
+// -1 for invalid characters
+static const int8_t HEX_TO_INT[256] = {
+    -1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,
+    -1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,
+    -1,-1, 0,1,2,3,4,5,6,7,8,9,-1,-1,-1,-1,-1,-1,-1,10,11,12,13,14,15,-1,
+    -1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,
+    -1,-1,10,11,12,13,14,15,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,
+    -1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,
+    -1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,
+    -1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,
+    -1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,
+    -1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,
+    -1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1
+};
+
+static inline void
+byte_to_hex(uint8_t byte, char *hex)
+{
+    hex[0] = INT_TO_HEX[(byte >> 4) & 0xf];
+    hex[1] = INT_TO_HEX[byte & 0xf];
+}
+
+static int
+from_hex(uuidobject *self, PyObject *hex)
```

Comment:
Do we really need this? can't we call the methods directly? I feel this overcomplicate the code. How much would we lose in terms of performance if we were to write in C what we currently do in Python for this part?

Reply:
Call which methods directly?  We do lose a lot in performance if we use Python, this code makes instantiation from string 7x faster. I'm not dropping this.

Reply:
What I meant is using PyObject_Call and call the appropriate methods instrad of reimplementing them in C. 

Also we maybe already have a str2int converter somewherr considering we have _Py_strhex(). Also if I would really prefer that we do not duplicate fromhex implementation when possible or at least, we should indicate that any changes in int.from_hex should be copied here as well.